File: packages/ruby/lib/nats_pubsub/subscribers/error_handler.rb
# frozen_string_literal: true

require_relative '../core/error_action'
require_relative '../core/constants'

module NatsPubsub
  module Subscribers
    # Enhanced error handler with ErrorAction support
    # Integrates per-subscriber error handling strategies
    class ErrorHandler
      attr_reader :subscriber, :config

      def initialize(subscriber, config: NatsPubsub.config)
        @subscriber = subscriber
        @config = config
      end

      # Handle an error from message processing
      #
      # @param error [Exception] The error that occurred
      # @param message [Hash] The message payload
      # @param context [MessageContext] Message metadata
      # @param attempt_number [Integer] Current delivery attempt
      # @return [Symbol] Error action to take (:retry, :discard, :dlq)
      def handle_error(error, message, context, attempt_number)
        error_context = build_error_context(error, message, context, attempt_number)

        # Try subscriber's custom error handler first
        action = if subscriber.respond_to?(:on_error)
                   call_subscriber_error_handler(error_context)
                 else
                   determine_default_action(error_context)
                 end

        # Validate and normalize action
        normalize_action(action, error_context)
      end

      private

      # Build error context object
      def build_error_context(error, message, context, attempt_number)
        Core::ErrorContext.new(
          error: error,
          message: message,
          context: context,
          attempt_number: attempt_number,
          max_attempts: config.max_deliver
        )
      end

      # Call subscriber's custom error handler
      def call_subscriber_error_handler(error_context)
        subscriber.on_error(error_context)
      rescue StandardError => e
        # If custom handler fails, log and use default
        config.logger&.error(
          "Subscriber error handler failed: #{e.class} - #{e.message}",
          subscriber: subscriber.class.name,
          error: e.class.name
        )
        determine_default_action(error_context)
      end

      # Determine default error action based on error type
      def determine_default_action(error_context)
        error = error_context.error

        # Malformed messages -> Discard immediately
        return Core::ErrorAction::DISCARD if malformed_error?(error)

        # Unrecoverable errors -> DLQ immediately
        return Core::ErrorAction::DLQ if unrecoverable_error?(error)

        # Last attempt -> DLQ
        return Core::ErrorAction::DLQ if error_context.last_attempt?

        # Default -> Retry with backoff
        Core::ErrorAction::RETRY
      end

      # Check if error indicates malformed message
      def malformed_error?(error)
        Constants::Errors::MALFORMED_ERRORS.any? do |error_name|
          error.class.name.include?(error_name)
        end
      end

      # Check if error is unrecoverable
      def unrecoverable_error?(error)
        Constants::Errors::UNRECOVERABLE_ERRORS.any? do |error_name|
          error.class.name.include?(error_name)
        end
      end

      # Normalize action to valid ErrorAction constant
      def normalize_action(action, error_context)
        return action if Core::ErrorAction.valid?(action)

        config.logger&.warn(
          "Invalid error action returned: #{action.inspect}, using default",
          subscriber: subscriber.class.name,
          valid_actions: Core::ErrorAction::ALL
        )

        determine_default_action(error_context)
      end
    end
  end
end

File: packages/javascript/src/subscribers/error-handler.ts
import { ErrorAction, ErrorContext, MessageContext } from '../types';
import { Errors } from '../core/constants';
import config from '../core/config';

/**
 * Subscriber interface with optional error handler
 */
interface SubscriberWithErrorHandler {
  onError?(errorContext: ErrorContext): Promise<ErrorAction> | ErrorAction;
  [key: string]: unknown;
}

/**
 * Enhanced error handler with ErrorAction support
 * Integrates per-subscriber error handling strategies
 */
export class ErrorHandler {
  constructor(
    private readonly subscriber: SubscriberWithErrorHandler,
    private readonly cfg = config.get()
  ) {}

  /**
   * Handle an error from message processing
   *
   * @param error The error that occurred
   * @param message The message payload
   * @param context Message metadata
   * @param attemptNumber Current delivery attempt
   * @returns Error action to take (RETRY, DISCARD, DLQ)
   */
  async handleError(
    error: Error,
    message: Record<string, unknown>,
    context: MessageContext,
    attemptNumber: number
  ): Promise<ErrorAction> {
    const errorContext = this.buildErrorContext(error, message, context, attemptNumber);

    // Try subscriber's custom error handler first
    let action: ErrorAction;
    if (this.subscriber.onError) {
      try {
        action = await this.subscriber.onError(errorContext);
      } catch (handlerError) {
        const logger = this.cfg.logger;
        if (logger && typeof logger.error === 'function') {
          logger.error(`Subscriber error handler failed: ${handlerError}`, {
            subscriber: this.subscriber.constructor.name,
            error: handlerError,
          });
        }
        action = this.determineDefaultAction(errorContext);
      }
    } else {
      action = this.determineDefaultAction(errorContext);
    }

    return this.normalizeAction(action, errorContext);
  }

  /**
   * Build error context object
   */
  private buildErrorContext(
    error: Error,
    message: Record<string, unknown>,
    context: MessageContext,
    attemptNumber: number
  ): ErrorContext {
    const maxAttempts = this.cfg.maxDeliver || 5;
    return {
      error,
      errorClass: error.constructor.name,
      errorMessage: error.message,
      message,
      metadata: context,
      attemptNumber,
      maxAttempts,
      lastAttempt: attemptNumber >= maxAttempts,
      remainingAttempts: Math.max(maxAttempts - attemptNumber, 0),
      retryable: this.isRetryableError(error),
    };
  }

  /**
   * Determine default action based on error type
   */
  private determineDefaultAction(errorContext: ErrorContext): ErrorAction {
    const { error, lastAttempt } = errorContext;

    // Malformed messages -> Discard immediately
    if (this.isMalformedError(error)) {
      return ErrorAction.DISCARD;
    }

    // Unrecoverable errors -> DLQ immediately
    if (this.isUnrecoverableError(error)) {
      return ErrorAction.DLQ;
    }

    // Last attempt -> DLQ
    if (lastAttempt) {
      return ErrorAction.DLQ;
    }

    // Default -> Retry
    return ErrorAction.RETRY;
  }

  /**
   * Check if error indicates malformed message
   */
  private isMalformedError(error: Error): boolean {
    return Errors.MALFORMED_ERRORS.some((errorName) => error.constructor.name.includes(errorName));
  }

  /**
   * Check if error is unrecoverable
   */
  private isUnrecoverableError(error: Error): boolean {
    return Errors.UNRECOVERABLE_ERRORS.some((errorName) =>
      error.constructor.name.includes(errorName)
    );
  }

  /**
   * Check if error is retryable
   */
  private isRetryableError(error: Error): boolean {
    return Errors.TRANSIENT_ERRORS.some((errorName) => error.constructor.name.includes(errorName));
  }

  /**
   * Normalize action to valid ErrorAction
   */
  private normalizeAction(action: ErrorAction, errorContext: ErrorContext): ErrorAction {
    const validActions = Object.values(ErrorAction);
    if (validActions.includes(action)) {
      return action;
    }

    const logger = this.cfg.logger;
    if (logger && typeof logger.warn === 'function') {
      logger.warn(`Invalid error action returned: ${action}, using default`, {
        subscriber: this.subscriber.constructor.name,
        validActions,
      });
    }

    return this.determineDefaultAction(errorContext);
  }
}
